### ISSUE TITLE
Print work order with latest improvements now prints, but only Labels for both Paperwork and labels.

### ISSUE THREAD

**Issue by efeoli** [timschofield/webERP#370]
Hi All,
title is descriptive of the issue.

Enrique

**Comment by efeoli** [comment]
Hi Tim,
clarifying this issue, you get either by clicking either the Paperwork or the Labels tab, labels as shown below, and clicking Paperwork.

![Image](https://github.com/user-attachments/assets/7a7e00e4-f33d-4dd8-9421-2ac361a41d40)


Enrique

**Comment by efeoli** [comment]
Hi Tim,
if this printing problem could be fixed, webERP could be in my opinion, very close to production status.

Enrique

**Comment by efeoli** [comment]
Hi Tim,
this bug is about the only serious obstacle to fully start using 5.00 22+rc in a production environment. For users that use wiki integration with webERP it is getting urgent to migrate to PHP 8 or above, since Dokuwiki, in my instance, now has features not working with PHP7. May I ask you kindly to please address this fix when you will have a free moment. Thank you for what all you and other developers do for this remarkable ERP.

Regards
Enrique

**Comment by dalers** [comment]
Hi @efeoli, can you please expand on how you are using webERP now and what your final goal is?

Iiuc you are running webERP and Dokuwiki on the same server in PHP7 and need PHP 8 for new Dokuwiki features, which requires that webERP be PHP8 compatible - i.e. v5 release. Do I understand correctly?

Fwiw, I'm using WackoWiki, I switched from MediaWiki to WackoWiki 1-2 yrs ago because (iiuc) I was forced to upgrade my server to PHP 8 for other reasons and WackoWiki was the only wiki supported by webERP that claimed PHP 8 compatibility.

**Comment by efeoli** [comment]
Thank you for your response,

Yes I am running Dokuwiki with webERP on the same server and same subdomain. Dokuwiki runs better in PHP 8 than in PHP 7, so I presume that although not officially claimed, it is set for PHP 8. Also, some plugins in Dokuwiki do not work with PHP 7. Therefore v%5is getting to be a necessity.

Regarding WackoWiki switch, it would be problematic for me since in Dokuwiki we have accumulated best practice norms and master files in sizable proportions that I imagine would limit migration.

I have started cautiously V5 with Dokuwiki and mostly both run quite fast and well. Dokuwiki runs faster in PHP 8 compared to PHP 7. Besides the bug mentioned above, I have found that work orders cannot be edited once they are entered (#503). Sometimes it is necessary to finetune quantities and this cannot be done. If these 2 issues could be solved, I think v5 would be acceptable for production environment while the resting issues get fixed and I see they aren't too many.

Enrique

**Comment by dalers** [comment]
Hi Enrique, 

Thank you for explaining your environment. webERP is complex software and everyone benefits from knowing what features others are using and finding valuable.

> I have found that work orders cannot be edited once they are entered ([#503](https://github.com/timschofield/webERP/issues/503)). Sometimes it is necessary to finetune quantities and this cannot be done.

Thank you for taking the time to create the issue!

Cheers,
Dale

**Comment by efeoli** [comment]
Hi Dale and Tim,
what Dale shows #338 for entering work orders persists, somewhat differently. You can't enter the order, you get this notice: 
**Fatal error**: Uncaught Error: The script tried to modify a property on an incomplete object. Please ensure that the class definition "WorkOrder" of the object you are trying to operate on was loaded _before_ unserialize() gets called or provide an autoloader to load the class definition in /home/bioacylc/domains/bioacyl.com/erp.bioacyl.com/webERP/WorkOrderEntry.php:36 Stack trace: #0 {main} thrown in /home/bioacylc/domains/bioacyl.com/erp.bioacyl.com/webERP/WorkOrderEntry.php on line 36 

This is with latest's commits since an hour ago.

Work orders now print, but quite differently for  v 5.0.0 rc 23 in the Demo.
Details are missing in the header and QR codes in the footer. Please see attachment.

[bioacylc_erp2_WorkOrder_656_2025-08-27.pdf](https://github.com/user-attachments/files/22014408/bioacylc_erp2_WorkOrder_656_2025-08-27.pdf)

**Comment by timschofield** [comment]
Hi @gggeek the recent mass changes you have done seem to have broken all the scripts that use our classes, not just work order entry. Can you have a look at this?

Tim

**Comment by timschofield** [comment]
> Hi [@gggeek](https://github.com/gggeek) the recent mass changes you have done seem to have broken all the scripts that use our classes, not just work order entry. Can you have a look at this?
> 
> Tim

Ah, looking at this the class definition needs to be included before session.php.

**Comment by gggeek** [comment]
@timschofield which classes in particular ?

**Comment by gggeek** [comment]
it seems this stems from saving php objects into the session. Hopefully it is only happening in WorkOrderEntry.php

**Comment by timschofield** [comment]
This one is the class WorkOrder defined in includes/DefineWOClass.php, but I have tried stock adjustments and sales orders as well and they are all broken.

**Comment by gggeek** [comment]
:-(

Putting php objects in the session is not a pattern I was aware of.

I have sent an initial fix PR: #653

**Comment by gggeek** [comment]
It seems there might be 10 files affected

**Comment by gggeek** [comment]
working on fixing all other cases...

**Comment by gggeek** [comment]
Final fixes in #654

**Comment by gggeek** [comment]
Sorry for the breackage.

For the record, I was looking into sending an http GET request to all webERP pages, as part of the CI tests happening on github, in order to prevent errors such as this.
Sadly, I do not think that this would have been enough to catch this issue, as, in order to put objects in the session, most likely the user needs to have some interaction with the app first...

**Comment by timschofield** [comment]
Thanks for sorting @gggeek 

It's one of the reasons for my allergy to big pull requests. I should have picked it up on review but the PR was big and I missed it :)

Tim

**Comment by gggeek** [comment]
> It's one of the reasons for my allergy to big pull requests. I should have picked it up on review but the PR was big and I missed it :)

I agree big PRs should not be a common thing.

To be fair, in this case it was not even one, but a series of PRs.

And for _some_ of the modified web pages, I was more cautious, and left the custom classes includes at their position. But I got bolder as I progressed, and thought I had properly checked for no side effects to class inclusion - but totally missed the session deserialisation issue...

**Comment by efeoli** [comment]
Hi All,
this fix 

<img width="276" height="64" alt="Image" src="https://github.com/user-attachments/assets/19f9a560-f798-484d-a82b-db5012ca57b5" />

has not corrected the issue.

**Comment by timschofield** [comment]
Hi Enrique, are you saying you still can't create a work order? I have just tried on both my local system, and on the oinline demo and it seems to work.

Tim

**Comment by efeoli** [comment]
Hi Tim,
yes, The demo seems to be one commit behind in the database version. It is 23 instead of 24. Is this right? In the 23 version I am able to create WO and they print correctly.

But with latest commit mounted on a fresh install and database inported from latest v 5.0.0 RC backup 24, I still get the error notice.

**Comment by timschofield** [comment]
> Hi Tim, yes, The demo seems to be one commit behind in the database version. It is 23 instead of 24. Is this right? In the 23 version I am able to create WO and they print correctly.

The demo is running 24 and I have just created work order 221 on it

<img width="498" height="95" alt="Image" src="https://github.com/user-attachments/assets/c0b94ef3-00a2-4a42-94b8-505ea8a8c16f" />

**Comment by efeoli** [comment]
For the demo, now everything works fine. The problem arises when I try to import the 4.15.2 database. Then, entering new order or introducing text to old orders generates the following warning:

<img width="1912" height="259" alt="Image" src="https://github.com/user-attachments/assets/97272430-92b0-4a8d-a12f-dfcbe5b6cfe5" />

All with a fresh install for the latest commit.

**Comment by timschofield** [comment]
You aren't using the latest version of WorkOrderEntry.php because line 36 is blank in that file - https://github.com/timschofield/webERP/blob/master/WorkOrderEntry.php

Tim

**Comment by efeoli** [comment]
I appreciate very much the tip Tim. Something must have failed during the installation. I tried again, and now it is working.

**Comment by timschofield** [comment]
Glad it's working for you now Enrique.

**Comment by gggeek** [comment]
So this can be closed, or is the original issue still lingering, and the only recent fix was to the extra breackage which I had piled on top?

--

Famous last words: if the code was using class autoloading, that breackage would not have happened 😉

**Comment by timschofield** [comment]
> Famous last words: if the code was using class autoloading, that breackage would not have happened 😉

Let me try and explain again. We don't code in the style that we do because we are thick or stupid. From the very start we wanted code that an accountant or other business person with an interest in coding could alter or add to. So if they wanted to do something they could load the script (it's easy to know which script, because our simple directory structure means they just have to look at the URL in their browser) in a text editor of their choice. They could quickly read the script from top to bottom, and where code is brought in from another file it is easy to follow the include. Making changes is also reasonably easy for these same reasons.
I get that for experienced and knowledgeable PHP programmers such as yourself, this seems odd, and probably just wrong. However I think if we changed to a system that you would find more to your liking then these people I mentioned would find the learning curve way too steep. In fact I probably wouldn't have joined the project myself. To be honest I am regretting getting as involved with composer as we are.
Your improvements to webERP are great, and I hope they continue, but I want to try and explain that there are reasons we do what we do, we don't just do it to annoy.

Thanks,
Tim

**Comment by gggeek** [comment]
> However I think if we changed to a system that you would find more to your liking then these people I mentioned would find the learning curve way too steep

This is possible.

Although, in my experience, most junior developers and non-formally-trained developers actually do like "automagic" a fair bit, and are happy to contribute by copy-pasting, even when they are not fully understanding what is going on.

As for the current system being easy to understand for newcomers, I am not so sure it is. It has a very flat folder structure, but the abundance of files makes it hard to navigate in any case. And the number of files included by every web page, including recursive includes, is beyond what one person can easily keep in his head.

I surmise that most business persons wanting to contribute to the project today are probably installing VSCode as first step, given its free-as-in-beer nature, and it should not be too hard for them to pick up basic code navigation skills.

> To be honest I am regretting getting as involved with composer as we are.

I am sorry that you feel that way, but agree that Composer, while being one of the best dependency managers out there, has its share of raw edges. In most projects I worked on, updating dependencies or running composer tasks other than `install` was mostly left to the lead developer.

That is compounded by the fact that shipping the `vendors` folder goes against the standard practice, and makes it not ok to just merge the PRs sent by google-bot.

Still, I think that there is a lot of value in using Composer to manage the dependencies, starting with being able to run security scanners on the lock file.

I have already added a script that checks that the composer dependencies and the lock files and autoload config are all in sync, in the build directory.

I have also added to my todo list to write a developer doc explaining the details of the way the project uses composer, how it differs from the standard and what to do or not to do, as well as one or two extra shell scripts to facilitate the merge of googlebot prs and updates of dependencies.

**Comment by timschofield** [comment]
> As for the current system being easy to understand for newcomers, I am not so sure it is. It has a very flat folder structure, but the abundance of files makes it hard to navigate in any case. And the number of files included by every web page, including recursive includes, is beyond what one person can easily keep in his head.

Well today I was looking at some issues with the MRP. If I hadn't already known it, then a quick look at the URL in my browser would have told me the file in question was MRP.php. One quick command later I had opened that file in my favourite text editor. I could read the file easily, despite it being one of the most complex scripts that we have. I am now testing the fix for the issue. I'm not sure what the problem with that workflow is? No problem with the flat diractory structure, in fact it made it easier - no need to find which sub directory it was in. No need to search through external files to find the SQL queries, they were right there in the code where they are needed. No need to learn a new IDE, or any other tools. I'm a humble accountant, I like a simple life withour artificially created problems.

I sense we will never agree on this :)

Tim

**Comment by efeoli** [comment]
Hia All,
the WO printing refinements, from this morning, did a nice job, and the outcome is very much as in 4.15.2. 

Could the printing of labels be separated from the WO printing? The tab still generates the WO, per se.

Enrique

**Comment by efeoli** [comment]
Hi All,
commit [6068f7e](https://github.com/timschofield/webERP/commit/6068f7e5b5ef52836447b253299f38f9f56fa511) solved this problem. Now WO and Labels print correctly. I will close the issue.
Enrique

**Comment by timschofield** [comment]
Thanks Enrique, sorry it took so long but I got there in the end :)
